node/actors/executor/src/attestation.rs

```diff
@@ -0,0 +1,89 @@
+//! Module to publish attestations over batches.
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use zksync_concurrency::ctx;
+use zksync_concurrency::time;
+use zksync_consensus_network::gossip::BatchVotesPublisher;
+use zksync_consensus_roles::attester::BatchNumber;
+use zksync_consensus_storage::{BatchStore, BlockStore};
+
+use crate::Attester;
+
+const POLL_INTERVAL: time::Duration = time::Duration::seconds(1);
+
+/// Polls the database for new batches to be signed and publishes them to the gossip channel.
+pub(super) struct AttesterRunner {
+    block_store: Arc<BlockStore>,
+    batch_store: Arc<BatchStore>,
+    attester: Attester,
+    publisher: BatchVotesPublisher,
+}
+
+impl AttesterRunner {
+    /// Crete a new instance of a runner.
+    pub(super) fn new(
+        block_store: Arc<BlockStore>,
+        batch_store: Arc<BatchStore>,
+        attester: Attester,
+        publisher: BatchVotesPublisher,
+    ) -> Self {
+        Self {
+            block_store,
+            batch_store,
+            attester,
+            publisher,
+        }
+    }
+    /// Poll the database for new L1 batches and publish our signature over the batch.
+    pub(super) async fn run(self, ctx: &ctx::Ctx) -> ctx::Result<()> {
+        let public_key = self.attester.key.public();
+        // The first batch number we want to publish our vote for. We don't have to re-publish a vote
+        // because once it enters the vote register even future peers can pull it from there.
+        let mut min_batch_number = BatchNumber(0);
+        loop {
+            // Pretend that the attesters can evolve.
+            let Some(attesters) = self.block_store.genesis().attesters.as_ref() else {
+                continue;
+            };
+            if !attesters.contains(&public_key) {
+                continue;
+            }
+
+            let mut unsigned_batch_numbers = self
+                .batch_store
+                .unsigned_batch_numbers(ctx)
+                .await
+                .context("unsigned_batch_numbers")?;
+
+            // Just to be sure we go from smaller to higher batches.
+            unsigned_batch_numbers.sort();
+
+            for bn in unsigned_batch_numbers {
+                // If we have already voted on this we can move on, no need to fetch the payload again.
+                // Batches appear in the store in order, even if we might have QC for a newer and not for an older batch,
+                // so once published our vote for a certain height, we can expect that we only have to vote on newer batches.
+                if bn < min_batch_number {
+                    continue;
+                }
+
+                if let Some(batch) = self
+                    .batch_store
+                    .batch_to_sign(ctx, bn)
+                    .await
+                    .context("batch_to_sign")?
+                {
+                    min_batch_number = batch.number.next();
```

Comment:
So I understand we always expect the `unsigned_batch_numbers` to be consecutive - you also left a comment that 
> // Batches appear in the store in order, even if we might have QC for a newer and not for an older batch

This makes sense, but then the API and the overall loop flow seems hard to follow: if `unsigned_batch_numbers` is supposed to be consecutive, why do we represent it as an array, not range? Now on the type level it's possible to represent errounous sequence like `4, 5, 7`. 

Also, feels confusing to have both an array of values to process and a mutable variable `min_batch_number` that allows for filtering such sequence.

Could we maybe keep a mutable range instead and operate over it? And thus only request the `last_persisted_batch` from `batch_store` to update the upper limit. 

There is probably a good reason why it is made like it is but then we should question the unerlying APIs/design decisions

Reply:
Thanks for taking a close look at this; if it's confusing, I might have misunderstood something. 

> we always expect the `unsigned_batch_numbers` to be consecutive

Not necessarily. By saying "batches appear in order" I simply meant that the actual records in `l1_batches` will be be consecutive, we won't see batch 6 appear in that table before batch 5. However the result of `unsigned_batch_numbers` can have gaps in it, because the records in `l1_batches_consensus` can appear in arbitrary order, once we have the `l1_batch`. That's exactly why it's an array and not a range. 

Say the last finalized batch is batch 3, and we have batches [4,5,6,7,8] in the database, and just so it happens that our node has already collected and inserted batch QC for batch 6 and 8. When we query `unsigned_batch_numbers` it will return [4,5,7], as per your example. This just means that we are still interested in collecting signatures over 4, 5 and 7, but we don't need to sign 6 and 8 as we already have the QC. 

So, we start by retrieving the `Batch` with the `hash` for 4, sign it, publish it, and raise the `min_batch_number` to 4+1. We never have to sign 4 again until we restart. Then, we sign and publish 5, and finally 7. Next time we look for unsigned numbers, we can look from `min_block_number=8`, since everything that was before that we already signed. The query won't return 8 because it has a QC, but when 9 appears, we can sign it. 

The only thing that can throw this off at the moment is if `batch_to_sign` happened to return `None` for 4 and `Some` for 5, but I believe/hope that would make no sense, as it would mean we have the commitment for 5 earlier than 4, and I'm pretty sure that would break some recursive property of proofs.

You may be right that it's a code smell to query the `unsigned_batch_numbers` array and _then_ filter it. In fact in https://github.com/matter-labs/zksync-era/pull/2399 a very similar `min_batch_number` was passed to the query. It means something slightly different though: in the query its purpose is to represent the "last finalized batch", whatever that might be - currently it's just the last batch minus 10, but eventually it should come from the main node or L1. But from the outside perspective I thought the `PersistentBatchStore` (or the `BatchStore`, depending on where we want to put an L1 API client) itself is a self-contained authority on what needs to be signed. This particular filter is just a way to prevent repeatedly retrieving the payload from the DB to hash and sign it, when we have already just done this. Only this loop knows that, though. So it's just a separation of concerns based on who is supposed to possess what information.

Say we passed this filter into the `unsigned_batch_numbers` right here. The inner query might know what the latest finalized batch is (either because it's flagged in the DB, or there is an L1 client, or the main node has been queried), but now it doesn't know if it should override the value passed in by this loop, or not. That's why it's actually the maximum of two numbers: `where l1_batch.number > last_signed_by_us and l1_batch.number > last_finalized`.

Reply:
> Could we maybe keep a mutable range instead and operate over it? And thus only request the last_persisted_batch from batch_store to update the upper limit.

IIUC that would be something like using `PersistentBatchStore::last_batch` for the upper limit, and introducing a `PersistentBatchStore::first_unsigned_batch() -> BatchNumber` or `PersistentBatchStore::last_finalized_batch() -> Option<BatchNumber>`, and then signing everything in between, periodically asking whether the upper limit can move on to the next step. The difference would be that it wouldn't skip all the numbers in between for which we already have a QC, unless we query those explicitly as we go to skip them, or retrieve and sign the payload regardless. 

I don't have a strong opinion on which one is better. For sure the main node API will need a `last_finalized_batch` query, which we discussed is one that doesn't have a gap after it. Implementing it now will mean we can have 1 less query methods.

Reply:
I think another reason I went with this because it deferred most of the logic to the store, and I couldn't really come up with a good definition for a `last_finalized_batch` method. Does it mean it's on L1 and L1 is final? Surely just because it's signed doesn't actually mean it's "final". So last "quorumed"? 

How about `get_earliest_unsigned_batch_number`, returning `Some` or `None`? Not as nice as something that would return a finalized height, but it might be the only thing we can do. 

And then it's just extended to return all of them, but I do see the appeal of not repeatedly returning 10 items just to discard all of them, but simply querying the `last_batch`, which saves a few roundtrips in all except the first loop.

If you agree with this "earliest unsigned" I'll rewrite the two PRs in terms of that.

Reply:
hey thanks for all this context! It makes sense and I understand the reasoning. I marginally prefer `get_earliest_unsigned_batch_number` over the current implementation, but also feel free to keep as is - we may revisit all of this after we have the first testing results.

Reply:
Okay, let's rewrite. It will save an extra query because the current way of determining the minimum height to consider for signing requires querying the maximum first.

Reply:
https://github.com/matter-labs/era-consensus/pull/150